litestar/plugins/flash.py

```diff
@@ -46,6 +48,8 @@ def on_app_init(self, app_config: AppConfig) -> AppConfig:
         Returns:
             The application configuration with the message callable registered.
         """
+        if self.config.session_config is None:
+            raise litestar.exceptions.ImproperlyConfiguredException("Flash messages require a session middleware.")
```

Comment:
```suggestion
        for mw in app_config.middleware:
            if isinstance(mw, SessionMiddleware):
                break
        else:
            raise litestar.exceptions.ImproperlyConfiguredException("Flash messages require a session middleware.")
```
My thinking here is that we won't need the session config object attached to the flash config.

Is it important that the messages are stored in a backend session?

Reply:
doing this change breaks the tests, the ImproperlyConfiguredException is always raised because I think by the time the plugin initializes the app doesnt know about middlewares yet. is that possible ?

Reply:
I think it is important messages are stored in whatever survives the orginal request scope, so maybe a backend session is only one of the options I dont know, I tried a little with context vars but while this passes the test it's because it keeps adding on the messages so the tests are not that good if not paying attention (https://github.com/litestar-org/litestar/pull/3420/commits/fd55270dd9a80857c2f8f9d62571a8dca069734b)

Reply:
> doing this change breaks the tests, the ImproperlyConfiguredException is always raised because I think by the time the plugin initializes the app doesnt know about middlewares yet. is that possible ?

I should have checked, the session config adds an instance of `DefineMiddleware`, so it would need to be `if isinstance(mw, DefineMiddleware) and isinstance(mw.middleware, SessionMiddleware):`

Reply:
> I think it is important messages are stored in whatever survives the orginal request scope, so maybe a backend session is only one of the options I dont know, I tried a little with context vars but while this passes the test it's because it keeps adding on the messages so the tests are not that good if not paying attention (https://github.com/litestar-org/litestar/pull/3420/commits/fd55270dd9a80857c2f8f9d62571a8dca069734b)

Sorry, my question should have been more clear. Is there a reason that a client side session is unsuitable? By checking for presence of a session middleware, it could be configured either with client side or server side storage.

Reply:
> if isinstance(mw, DefineMiddleware) and isinstance(mw.middleware, S

ok this is weird, the tests fail with this:
` isinstance(mw, DefineMiddleware) ` is True 
but ` isinstance(mw.middleware, SessionMiddleware)` is False

Reply:
> > I think it is important messages are stored in whatever survives the orginal request scope, so maybe a backend session is only one of the options I dont know, I tried a little with context vars but while this passes the test it's because it keeps adding on the messages so the tests are not that good if not paying attention ([fd55270](https://github.com/litestar-org/litestar/commit/fd55270dd9a80857c2f8f9d62571a8dca069734b))
> 
> Sorry, my question should have been more clear. Is there a reason that a client side session is unsuitable? By checking for presence of a session middleware, it could be configured either with client side or server side storage.

I dont think there is any importance

Reply:
would `issubclass(mw.middleware, SessionMiddleware)` be ok you think @peterschutt ?

Reply:
Sorry, yeah my bad:

`if isinstance(mw, DefineMiddleware) and issubclass(mw.middleware, SessionMiddleware):`

Reply:
this pass tests with it but I truly dont get how to fix that mypy error

Reply:
B/c `DefineMiddleware.middleware` is typed as `Callable[..., ASGIApp]`, it doesn't have to be a class, but `issubclass()` will error if the thing we pass it isn't a class. So we have to check for that first. For this we have `litestar.utils.is_class_and_subclass()`.

I've just pushed a commit.

Reply:
I've found there is one other case where session middleware can exist in the middleware stack without us detecting here: if a user has added `SessionAuth` to the application. Adding that results in a type of `security.session_auth.middleware.MiddlewareWrapper` middleware added to the stack, inside of `DefineMiddleware`. And inside `MiddlewareWrapper` it creates its own little asgi middleware stack that include `SessionMiddleware` :face_with_spiral_eyes: 

So, i think we have to do:
```py
if isinstance(mw, DefineMiddleware) and issubclass(mw.middleware, (SessionMiddleware, MiddlewareWrapper)):
```
... to actually ascertain if `SessionMiddleware` is in the stack.

I think `SessionAuth` should behave exactly like you are implementing here, that is, not add session middleware to the stack, but require it to be there already.

I created a development thread for this in discord, here: https://discord.com/channels/919193495116337154/1232602408119570473

Reply:
I just added a test to up the coverage which dropped, this should be ok now I think

Reply:
> I've found there is one other case where session middleware can exist in the middleware stack without us detecting here: if a user has added `SessionAuth` to the application.
> 
> That results in `security.session_auth.middleware.MiddlewareWrapper` added to the middleware stack, itself wrapped inside of `DefineMiddleware`.
> 
> Inside `MiddlewareWrapper` it creates its own little asgi middleware stack that includes `SessionMiddleware` 😵‍💫
> 
> So, i think we have to do:
> 
> ```python
> if isinstance(mw, DefineMiddleware) and issubclass(mw.middleware, (SessionMiddleware, MiddlewareWrapper)):
> ```
> 
> ... to actually ascertain if `SessionMiddleware` is in the stack.
> 
> I think `SessionAuth` should behave exactly like you are implementing here, that is, not add session middleware to the stack, but require it to be there already.
> 
> I created a development thread for this in discord, here: https://discord.com/channels/919193495116337154/1232602408119570473

Added a commit for this - we should be good to go now!